File: src/dotnet/PloyBus.Tests/Transport/InMemory/InMemoryTransportTests.cs
using System.Diagnostics;
using Microsoft.Extensions.Logging;
using NUnit.Framework;

namespace PolyBus.Transport.InMemory;

[TestFixture]
class InMemoryTransportTests
{
    private TestEnvironment _testEnvironment;

    [SetUp]
    public async Task SetUp()
    {
        _testEnvironment = new();
        await _testEnvironment.Setup();
    }

    [TearDown]
    public async Task TearDown()
    {
        await _testEnvironment.Stop();
    }

    [Test, CancelAfter(5 * 60 * 1000)]
    public async Task Send_BeforeStarting()
    {
        // Arrange
        var transaction = await _testEnvironment.Beta.Bus.CreateOutgoingTransaction();
        var taskCompletionSource = new TaskCompletionSource<bool>();
        _testEnvironment.Alpha.OnMessageReceived = _ =>
        {
            // This should not be called
            taskCompletionSource.SetResult(true);
            return Task.CompletedTask;
        };

        // Act
        transaction.Add(new AlphaCommand { Name = "Test" });

        // Assert - should throw an error because the transport is not started
        Assert.ThrowsAsync<PolyBusNotStartedError>(transaction.Commit);
        Assert.That(taskCompletionSource.Task.IsCompleted, Is.False);
    }

    [Test, CancelAfter(5 * 60 * 1000)]
    public async Task Send_AfterStarted()
    {
        // Arrange
        var transaction = await _testEnvironment.Beta.Bus.CreateOutgoingTransaction();
        var taskCompletionSource = new TaskCompletionSource<bool>();
        _testEnvironment.Alpha.OnMessageReceived = _ =>
        {
            taskCompletionSource.SetResult(true);
            return Task.CompletedTask;
        };

        // Act - send a command from the beta endpoint to alpha endpoint
        await _testEnvironment.Start();
        transaction.Add(new AlphaCommand { Name = "Test" });
        await transaction.Commit();
        await taskCompletionSource.Task;

        // Assert
        Assert.That(taskCompletionSource.Task.IsCompleted, Is.True);
    }

    [Test, CancelAfter(5 * 60 * 1000)]
    public async Task Send_WithExplicitEndpoint()
    {
        // Arrange
        var transaction = await _testEnvironment.Alpha.Bus.CreateOutgoingTransaction();
        var taskCompletionSource = new TaskCompletionSource<string>();
        _testEnvironment.Alpha.OnMessageReceived = arg =>
        {
            // This should NOT be called
            taskCompletionSource.SetResult(arg.Bus.Name);
            return Task.CompletedTask;
        };
        _testEnvironment.Alpha.Transport.DeadLetterHandler = _ =>
        {
            // This should be called
            taskCompletionSource.SetResult(_testEnvironment.Alpha.Transport.DeadLetterEndpoint);
        };
        var endpoint = _testEnvironment.Alpha.Transport.DeadLetterEndpoint;

        // Act - send the alpha command to beta endpoint
        await _testEnvironment.Start();
        transaction.Add(new AlphaCommand { Name = "Test" }, endpoint: endpoint);
        await transaction.Commit();
        var actualEndpoint = await taskCompletionSource.Task;

        // Assert
        Assert.That(actualEndpoint, Is.EqualTo(endpoint));
    }

    [Test, CancelAfter(5 * 60 * 1000)]
    public async Task Send_WithHeaders()
    {
        // Arrange
        const string headerKey = "X-Custom-Header";
        const string headerValue = "HeaderValue";
        var transaction = await _testEnvironment.Alpha.Bus.CreateOutgoingTransaction();
        var taskCompletionSource = new TaskCompletionSource<string>();
        _testEnvironment.Alpha.OnMessageReceived = message =>
        {
            taskCompletionSource.SetResult(
                message.IncomingMessage.Headers.TryGetValue(headerKey, out var value)
                ? value
                : string.Empty);
            return Task.CompletedTask;
        };

        // Act - send a command with a custom header
        await _testEnvironment.Start();
        var message = transaction.Add(new AlphaCommand { Name = "Test" });
        message.Headers.Add(headerKey, headerValue);
        await transaction.Commit();
        var actualHeaderValue = await taskCompletionSource.Task;

        // Assert
        Assert.That(actualHeaderValue, Is.EqualTo(headerValue));
    }

    [Test, CancelAfter(5 * 60 * 1000)]
    public async Task Send_WithDelay()
    {
        // Arrange
        const int delay = 5000; // 5 seconds
        var transaction = await _testEnvironment.Alpha.Bus.CreateOutgoingTransaction();
        var stopwatch = new Stopwatch();
        var taskCompletionSource = new TaskCompletionSource<TimeSpan>();
        _testEnvironment.Alpha.OnMessageReceived = _ =>
        {
            stopwatch.Stop();
            taskCompletionSource.SetResult(stopwatch.Elapsed);
            return Task.CompletedTask;
        };

        // Act - send to the dead letters queue instead of normal processing queue
        await _testEnvironment.Start();
        var message = transaction.Add(new AlphaCommand { Name = "Test" });
        message.DeliverAt = DateTime.UtcNow.AddMilliseconds(delay);
        stopwatch.Start();
        await transaction.Commit();
        var elapsed = await taskCompletionSource.Task;
        _testEnvironment.InMemoryMessageBroker.Log.LogInformation("Elapsed time for delayed message: {Elapsed}",
            Math.Floor(elapsed.TotalMilliseconds));

        // Assert
        Assert.That(elapsed, Is
            .LessThanOrEqualTo(TimeSpan.FromMilliseconds(delay + 1000)) // allow 1 second of leeway
            .And.GreaterThanOrEqualTo(TimeSpan.FromMilliseconds(delay - 1000)));// allow 1 second of leeway
    }

    [Test, CancelAfter(5 * 60 * 1000)]
    public async Task Send_WithExpiredDelay()
    {
        // Arrange
        var transaction = await _testEnvironment.Alpha.Bus.CreateOutgoingTransaction();
        var taskCompletionSource = new TaskCompletionSource<bool>();
        _testEnvironment.Alpha.OnMessageReceived = _ =>
        {
            taskCompletionSource.SetResult(true);
            return Task.CompletedTask;
        };

        // Act - schedule command to be delivered in the past
        await _testEnvironment.Start();
        var message = transaction.Add(new AlphaCommand { Name = "Test" });
        message.DeliverAt = DateTime.UtcNow.AddMilliseconds(-1000); // 1 second in the past
        await transaction.Commit();
        await taskCompletionSource.Task;

        // Assert
        Assert.That(taskCompletionSource.Task.IsCompleted, Is.True);
    }

    [Test, CancelAfter(5 * 60 * 1000)]
    public async Task Start_WhenAlreadyStarted()
    {
        // Act
        await _testEnvironment.Start();

        // Assert - starting again should not throw an error
        Assert.DoesNotThrowAsync(_testEnvironment.Start);
    }

    [Test, CancelAfter(5 * 60 * 1000)]
    public async Task Subscribe_BeforeStarted()
    {
        // Arrange
        var transaction = await _testEnvironment.Alpha.Bus.CreateOutgoingTransaction();
        var taskCompletionSource = new TaskCompletionSource<bool>();
        _testEnvironment.Beta.OnMessageReceived = _ =>
        {
            taskCompletionSource.SetResult(true);
            return Task.CompletedTask;
        };

        // Act - subscribing before starting should throw an error
        Assert.ThrowsAsync<PolyBusNotStartedError>(async () =>
        {
            await _testEnvironment.Beta.Transport.Subscribe(
                _testEnvironment.Beta.Bus.Messages.GetMessageInfo(typeof(AlphaEvent))!);
        });
        transaction.Add(new AlphaEvent { Name = "Test" });
        Assert.ThrowsAsync<PolyBusNotStartedError>(transaction.Commit);

        // Assert
        Assert.That(taskCompletionSource.Task.IsCompleted, Is.False);
    }

    [Test, CancelAfter(5 * 60 * 1000)]
    public async Task Subscribe()
    {
        // Arrange
        var transaction = await _testEnvironment.Alpha.Bus.CreateOutgoingTransaction();
        var taskCompletionSource = new TaskCompletionSource<bool>();
        _testEnvironment.Beta.OnMessageReceived = _ =>
        {
            taskCompletionSource.SetResult(true);
            return Task.CompletedTask;
        };
        await _testEnvironment.Start();

        // Act - subscribing before starting should throw an error
        await _testEnvironment.Beta.Transport.Subscribe(
            _testEnvironment.Beta.Bus.Messages.GetMessageInfo(typeof(AlphaEvent))!);
        transaction.Add(new AlphaEvent { Name = "Test" });
        await transaction.Commit();
        await taskCompletionSource.Task;
    }
}

File: src/kotlin/src/test/kotlin/polybus/transport/inmemory/InMemoryTransportTests.kt
package polybus.transport.inmemory

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.fail
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import polybus.transport.PolyBusNotStartedError
import java.time.Instant

class InMemoryTransportTests {
    private lateinit var testEnvironment: TestEnvironment

    @BeforeEach
    fun setUp() = runBlocking {
        testEnvironment = TestEnvironment()
        testEnvironment.setup()
    }

    @AfterEach
    fun tearDown() = runBlocking {
        testEnvironment.stop()
    }

    @Test
    fun sendBeforeStarting() = runBlocking {
        // Arrange
        val transaction = testEnvironment.beta.bus.createOutgoingTransaction()
        val called = CompletableDeferred<Boolean>()
        testEnvironment.alpha.onMessageReceived = { called.complete(true) }

        // Act
        transaction.add(AlphaCommand().apply { name = "Test" })

        // Assert - should throw an error because the transport is not started
        try {
            transaction.commit()
            fail("Expected PolyBusNotStartedError to be thrown")
        } catch (e: PolyBusNotStartedError) {
            // Expected - exception was thrown as expected
        }
        assertFalse(called.isCompleted)
    }

    @Test
    fun sendAfterStarted() = runBlocking {
        // Arrange
        val transaction = testEnvironment.beta.bus.createOutgoingTransaction()
        val called = CompletableDeferred<Boolean>()
        testEnvironment.alpha.onMessageReceived = { called.complete(true) }

        // Act - send a command from the beta endpoint to alpha endpoint
        testEnvironment.start()
        transaction.add(AlphaCommand().apply { name = "Test" })
        transaction.commit()

        // Assert
        assertTrue(called.await())
    }

    @Test
    fun sendWithExplicitEndpoint() = runBlocking {
        // Arrange
        val transaction = testEnvironment.alpha.bus.createOutgoingTransaction()
        val result = CompletableDeferred<String>()
        testEnvironment.alpha.onMessageReceived = {
            // This should NOT be called
            result.complete(it.bus.name)
        }
        testEnvironment.alpha.transport.deadLetterHandler = {
            // This should be called
            result.complete(testEnvironment.alpha.transport.deadLetterEndpoint)
        }
        val endpoint = testEnvironment.alpha.transport.deadLetterEndpoint

        // Act - send the alpha command to dead letter endpoint
        testEnvironment.start()
        transaction.add(AlphaCommand().apply { name = "Test" }, endpoint = endpoint)
        transaction.commit()

        // Assert
        assertEquals(endpoint, result.await())
    }

    @Test
    fun sendWithHeaders() = runBlocking {
        // Arrange
        val key = "X-Custom-Header"
        val value = "HeaderValue"
        val transaction = testEnvironment.alpha.bus.createOutgoingTransaction()
        val result = CompletableDeferred<String>()
        testEnvironment.alpha.onMessageReceived = {
            result.complete(it.incomingMessage.headers[key].orEmpty())
        }

        // Act - send a command with a custom header
        testEnvironment.start()
        val message = transaction.add(AlphaCommand().apply { name = "Test" })
        message.headers[key] = value
        transaction.commit()

        // Assert
        assertEquals(value, result.await())
    }

    @Test
    fun sendWithDelay() = runBlocking {
        // Arrange
        val delayMs = 5000L // 5 seconds
        val transaction = testEnvironment.alpha.bus.createOutgoingTransaction()
        val result = CompletableDeferred<Long>()
        var startedAt = 0L
        testEnvironment.alpha.onMessageReceived = {
            result.complete(System.currentTimeMillis() - startedAt)
        }

        // Act - send to the dead letters queue instead of normal processing queue
        testEnvironment.start()
        val message = transaction.add(AlphaCommand().apply { name = "Test" })
        message.deliverAt = Instant.now().plusMillis(delayMs)
        startedAt = System.currentTimeMillis()
        transaction.commit()
        val elapsedMs = result.await()

        // Assert - allow 1 second of leeway
        assertTrue(elapsedMs in (delayMs - 1000)..(delayMs + 1000))
    }

    @Test
    fun sendWithExpiredDelay() = runBlocking {
        // Arrange
        val transaction = testEnvironment.alpha.bus.createOutgoingTransaction()
        val called = CompletableDeferred<Boolean>()
        testEnvironment.alpha.onMessageReceived = {
            called.complete(true)
        }

        // Act - schedule command to be delivered in the past
        testEnvironment.start()
        val message = transaction.add(AlphaCommand().apply { name = "Test" })
        message.deliverAt = Instant.now().minusMillis(1000) // 1 second in the past
        transaction.commit()

        // Assert
        assertTrue(called.await())
    }

    @Test
    fun startWhenAlreadyStarted() = runBlocking {
        // Act
        testEnvironment.start()

        // Assert - starting again should not throw an error
        testEnvironment.start()
    }

    @Test
    fun subscribeBeforeStarted() = runBlocking {
        // Arrange
        val transaction = testEnvironment.alpha.bus.createOutgoingTransaction()
        val called = CompletableDeferred<Boolean>()
        testEnvironment.beta.onMessageReceived = { called.complete(true) }

        // Act + Assert - subscribing and sending before starting should throw
        try {
            testEnvironment.beta.transport.subscribe(
                testEnvironment.beta.bus.messages.getMessageInfo(AlphaEvent::class.java)
            )
            fail("Expected PolyBusNotStartedError to be thrown")
        } catch (e: PolyBusNotStartedError) {
            // Expected - exception was thrown as expected
        }
        transaction.add(AlphaEvent().apply { name = "Test" })
        try {
            transaction.commit()
            fail("Expected PolyBusNotStartedError to be thrown")
        } catch (e: PolyBusNotStartedError) {
            // Expected - exception was thrown as expected
        }

        // Assert
        assertFalse(called.isCompleted)
    }

    @Test
    fun subscribe() = runBlocking {
        // Arrange
        val transaction = testEnvironment.alpha.bus.createOutgoingTransaction()
        val called = CompletableDeferred<Boolean>()
        testEnvironment.beta.onMessageReceived = { called.complete(true) }

        // Act - subscribing before starting should throw an error
        testEnvironment.start()
        testEnvironment.beta.transport.subscribe(
            testEnvironment.beta.bus.messages.getMessageInfo(AlphaEvent::class.java)
        )
        transaction.add(AlphaEvent().apply { name = "Test" })
        transaction.commit()
        assertTrue(called.await())
    }
}
